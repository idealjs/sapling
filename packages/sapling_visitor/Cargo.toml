[package]
name = "sapling_visitor"
version = "0.1.0"
edition = "2024"
publish = false

[dependencies]
sapling_macros = { path = "../sapling_macros" }
sapling_shared = { path = "../sapling_shared" }
