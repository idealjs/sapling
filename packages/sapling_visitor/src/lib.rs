use indextree::{Arena, NodeId};
use oxc_allocator::Allocator;
use oxc_ast::AstKind;
use oxc_ast::ast::{Atom, Statement};
use oxc_ast_visit::walk_mut::walk_statement;
use oxc_ast_visit::{Visit, VisitMut};
use oxc_parser::Parser;
use oxc_span::SourceType;
use oxc_traverse::{Traverse, TraverseCtx};
pub struct Transformer<'a> {
    allocator: &'a Allocator,
}

impl<'a> Transformer<'a> {
    pub fn new(allocator: &'a Allocator) -> Self {
        Self { allocator }
    }
}

impl<'a> Transformer<'a> {}

impl<'a> Traverse<'a> for Transformer<'a> {
    fn enter_statement(&mut self, node: &mut Statement<'a>, ctx: &mut TraverseCtx<'a>) {}
}

pub struct SaplingVisitor<'a> {
    pub allocator: &'a Allocator,
}

impl<'a> VisitMut<'a> for SaplingVisitor<'a> {
    fn visit_statement(&mut self, node: &mut Statement<'a>) {
        walk_statement(self, node);
        if let Statement::FunctionDeclaration(func) = node {
            if let Some(name) = func.id.as_mut() {
                let uppercase_name = name.name.to_uppercase();
                let allocated = self.allocator.alloc_str(&uppercase_name);
                name.name = Atom::from(allocated);
            }
        }
    }
}
