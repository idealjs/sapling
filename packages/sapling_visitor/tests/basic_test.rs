use std::{fs, path::Path};

use oxc_allocator::Allocator;
use oxc_codegen::Codegen;
use oxc_parser::Parser;
use oxc_semantic::SemanticBuilder;
use oxc_span::SourceType;
use oxc_traverse::traverse_mut;
use sapling_visitor::{SaplingVisitor, Transformer};
use oxc_ast_visit::VisitMut; // Import the trait for visit_program

#[test]
fn test_uppercase_function_transform() {
    let path = Path::new("tests/fixtures/Test.tsx");
    let source_text = fs::read_to_string(path).unwrap();

    let allocator = Allocator::default();
    let source_type = SourceType::default().with_typescript(true).with_jsx(true);

    let ret = Parser::new(&allocator, &source_text, source_type).parse();
    let mut program = ret.program;

    let mut visitor = SaplingVisitor {
        allocator: &allocator,
    };
    
    visitor.visit_program(&mut program);

    let result = Codegen::new().build(&program);

    insta::assert_snapshot!(result.code);
}
