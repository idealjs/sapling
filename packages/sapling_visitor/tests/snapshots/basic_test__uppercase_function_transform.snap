---
source: packages/sapling_visitor/tests/basic_test.rs
expression: result.code
---
"use client";
const style = (values: unknown): string => {
	return "";
};
const Test = () => {
	return <div className={style({})}>Test</div>;
};
export default Test;
function SOMEFUNCTION() {
	return "This is a function";
}
